components/core/src/clp/ffi/Value.hpp

```diff
@@ -0,0 +1,224 @@
+#ifndef CLP_FFI_VALUE_HPP
+#define CLP_FFI_VALUE_HPP
+
+#include <cstdint>
+#include <string>
+#include <string_view>
+#include <tuple>
+#include <type_traits>
+#include <utility>
+#include <variant>
+
+#include "../ErrorCode.hpp"
+#include "../ir/EncodedTextAst.hpp"
+#include "../TraceableException.hpp"
+
+namespace clp::ffi {
+using value_int_t = int64_t;
+using value_float_t = double;
+using value_bool_t = bool;
+
+/**
+ * Tuple of all the valid primitive value types.
+ */
+using PrimitiveValueTypeTuple = std::tuple<
+        value_int_t,
+        value_float_t,
+        value_bool_t,
+        std::string,
+        clp::ir::EightByteEncodedTextAst,
+        clp::ir::FourByteEncodedTextAst>;
+
+/**
+ * Template that converts a tuple of types into a variant.
+ * @tparam Tuple A tuple of types
+ */
+template <typename Tuple>
+struct tuple_to_variant;
+
+template <typename... Types>
+struct tuple_to_variant<std::tuple<Types...>> {
+    using type = std::variant<std::monostate, Types...>;
+};
+
+/**
+ * Variant type defined by the tuple of all primitive value types.
+ */
+using PrimitiveValueTypeVariant = typename tuple_to_variant<PrimitiveValueTypeTuple>::type;
+
+/**
+ * Template to validate if the given type is in the type tuple.
+ * @tparam Type
+ * @tparam Tuple
+ */
+template <typename Type, typename Tuple>
+struct is_valid_type;
```

Comment:
* How about `is_in_type_tuple`?
* Like `tuple_to_variant`, can we move this into `type_utils.hpp`?